File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_MOVIMIENTOS_PREDEFINIDOS/P5-java/src/irrgarten/LabyrinthCharacter.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author ismael
 */
/**
 * La clase LabyrinthCharacter representa un personaje dentro de un laberinto.
 * Cada personaje tiene un nombre, inteligencia, salud, fuerza y una posición
 * específica en el laberinto definida por una fila y una columna.
 */
/**
 * Representa un personaje dentro de un laberinto.
 * 
 * <p>Esta clase contiene las características básicas de un personaje en el laberinto,
 * incluyendo su nombre, inteligencia, salud, fuerza y su posición en el laberinto.</p>
 * 
 * @author Ismael
 */
abstract class LabyrinthCharacter {
    /**
     * El nombre del personaje.
     */
    private String name;

    /**
     * La inteligencia del personaje, representada como un valor flotante.
     */
    private float intelligence;

    /**
     * La salud del personaje, representada como un valor flotante.
     */
    private float health;

    /**
     * La fuerza del personaje, representada como un valor flotante.
     */
    private float strength;

    /**
     * La fila en la que se encuentra el personaje dentro del laberinto.
     */
    private int row;

    /**
     * La columna en la que se encuentra el personaje dentro del laberinto.
     */
    private int col;

    /*
     * Constante definida para no usar números mágicos
     */
    private static final int INVALID_POS=-1;
    
    /**
     * Construye un nuevo LabyrinthCharacter con los atributos especificados.
     *
     * @param name el nombre del personaje
     * @param intelligence el nivel de inteligencia del personaje
     * @param strength el nivel de fuerza del personaje
     * @param health el nivel de salud del personaje
     */
    public LabyrinthCharacter(String name, float intelligence, float strength, float health){
        this.name = name;
        this.intelligence = intelligence;
        this.strength = strength;
        this.health = health;
        this.row = INVALID_POS;
        this.col = INVALID_POS;
    }

    /* 
     * Constructor de copia
     * @param objeto de tipo LabyrinthCharacter a copiar
     */
    public LabyrinthCharacter(LabyrinthCharacter other){
        this.name = other.name;
        this.intelligence = other.intelligence;
        this.strength = other.strength;
        this.health = other.health;
        this.row = other.row;
        this.col = other.col;
    }

    /**
 * Verifica si el objeto está muerto.
 * 
 * @return true si la salud es menor o igual a 0, de lo contrario false.
 */
public boolean dead(){
    return health <= 0;
}

/**
 * Obtiene la fila actual.
 * 
 * @return la fila actual.
 */
public int getRow(){
    return row;
}

/**
 * Obtiene la columna actual.
 * 
 * @return la columna actual.
 */
public int getCol(){
    return col;
}

/**
 * Obtiene la inteligencia.
 * 
 * @return la inteligencia.
 */
protected float getIntelligence(){
    return intelligence;
}

/**
 * Obtiene la salud.
 * 
 * @return la salud.
 */
protected float getHealth(){
    return health;
}

/**
 * Obtiene la fuerza.
 * 
 * @return la fuerza.
 */
protected float getStrength(){
    return strength;
}

/**
 * Establece la salud.
 * 
 * @param health la nueva salud.
 */
protected void setHealth(float health){
    this.health = health;
}

/**
 * Establece la posición.
 * 
 * @param row la nueva fila.
 * @param col la nueva columna.
 */
public void setPos(int row, int col){
    this.row = row;
    this.col = col;
}

/**
 * Devuelve una representación en cadena del objeto.
 * 
 * @return una cadena que representa el objeto.
 */
@Override
public String toString(){
    return name + " (" + row + ", " + col + ")" + "\n" + "Intelligence: " + intelligence + "\n" + "Strength: " + strength + "\n" + "Health: " + health;
}

/**
 * Reduce la salud en uno.
 */
protected void gotWounded(){
    health--;
}

/**
 * Realiza un ataque.
 * 
 * @return el valor del ataque.
 */
public abstract float attack();

/**
 * Defiende contra un ataque.
 * 
 * @param attack el valor del ataque recibido.
 * @return true si la defensa fue exitosa, de lo contrario false.
 */
public abstract boolean defend(float attack);
        
}

File: Asignaturas/Tercer Año/PDOO/Practica/Proyecto_Irrgarten/Proyecto_Irrgarten/SUBIR_A_PRADO_2_PARCIAL_PDOO/VARIOS_DISEÑOS/PROYECTO_SUPER_PLAYER/SolucionParcialPracticas2/P5-Ruby/lib/irrgarten/labyrinth_character.rb
#encoding: UTF-8
# @autor: Ismael Sallami Moreno
module Irrgarten
    class LabyrinthCharacter
        # La fila o columna inválida.
        @@INVALID_POS=-1
        
        # Initializes a new LabyrinthCharacter object.
        #
        # @param name [String] el nombre del personaje
        # @param intelligence [Integer] el nivel de inteligencia del personaje
        # @param strength [Integer] el nivel de fuerza del personaje
        # @param health [Integer] el nivel de salud del personaje
        def initialize(name, intelligence, strength, health)
            @name = name
            @intelligence = intelligence
            @strength = strength
            @health = health
            @row = @@INVALID_POS
            @col = @@INVALID_POS
        end


        #-----------definición de accessor, reader, ...-----------
        protected
        # Consultor de @name
        # @return [String] nombre del personaje
        attr_reader :name

        # Consultor de @intelligence
        # @return [float] inteligencia del personaje
        attr_reader :intelligence

        # Consultor de @strength
        # @return [float] fuerza del personaje
        attr_reader :strength

        # Consultor y modificador de @health
        # @return [float] salud del personaje
        attr_accessor :health

        public
        # Consultor de @row
        # @return [int] fila de la posición del personaje
        attr_reader :row

        # Consultor de @col
        # @return [int] columna de la posición del personaje
        attr_reader :col

        #-----------definición de accessor, reader, ...-----------


        # Modificador de la posición del personaje
        #
        # @param row [int] fila de la posición del personaje
        # @param col [int] columna de la posición del personaje
        def pos(row, col)
            @row = row
            @col = col
        end




        # Crea una copia de otro objeto LabyrinthCharacter.
        #
        # @param other [LabyrinthCharacter] el otro objeto LabyrinthCharacter
        # @return [LabyrinthCharacter] una nueva instancia de LabyrinthCharacter
        def copy(other) # copia superficial
            @name = other.name
            @intelligence = other.intelligence
            @strength = other.strength
            @health = other.health
            pos(other.row, other.col)
        end

        # Verifica si el personaje está muerto.
        #
        # @return [Boolean] true si el personaje está muerto, false en caso contrario
        def dead
            @health <= 0
        end

        # Obtiene la fila actual del personaje.
        #
        # @return [Integer] la fila actual del personaje
        def get_row
            @row
        end

        # Obtiene la columna actual del personaje.
        #
        # @return [Integer] la columna actual del personaje
        def get_col
            @col
        end

        protected 

        # Obtiene el nivel de inteligencia del personaje.
        #
        # @return [Integer] el nivel de inteligencia del personaje
        def get_intelligence
            @intelligence
        end

        # Obtiene el nivel de fuerza del personaje.
        #
        # @return [Integer] el nivel de fuerza del personaje
        def get_strength
            @strength
        end

        # Obtiene el nivel de salud del personaje.
        #
        # @return [Integer] el nivel de salud del personaje
        def get_health
            @health
        end

        # # Establece el nivel de salud del personaje.
        # #
        # # @param health [Integer] el nuevo nivel de salud del personaje
        # def set_health(health)
        #     @health = health
        # end

        public

        # Establece la posición del personaje en el laberinto.
        #
        # @param row [Integer] la nueva fila del personaje
        # @param col [Integer] la nueva columna del personaje
        def setPos(row, col)
            @row = row
            @col = col
        end

        # Devuelve una representación en cadena del personaje.
        #
        # @return [String] una cadena que representa al personaje
        def to_s
            "#{@name} (Intelligence: #{@intelligence}, Strength: #{@strength}, Health: #{@health}, Row: #{@row}, Col: #{@col})"
        end

        protected
        
        # Reduce la salud del personaje en 1.
        def gotWounded
            @health -= 1
        end

        public

        # Método abstracto para atacar.
        #
        # @raise [NotImplementedError] si el método no es implementado por una subclase
        def attack
            raise NotImplementedError
        end

        # Método abstracto para defenderse de un ataque.
        #
        # @param attack [Object] el ataque recibido
        # @raise [NotImplementedError] si el método no es implementado por una subclase
        def defend(attack)
            raise NotImplementedError
        end
    end
end




